File: backend-service/src/utils/response.ts
/* Imports */
import type { Context } from "hono";
import type { StatusCode } from "hono/utils/http-status";

/* Local Imports */
import { RESPONSE_MESSAGES, STATUS_CODES } from "@/constants/index.js";

// ----------------------------------------------------------------------

interface SuccessResponseData {
  status: {
    response_code: StatusCode;
    response_message: string;
  };
  message: any;
  data?: any;
}

interface ErrorResponseData {
  status: {
    response_code: StatusCode;
    response_message: string;
  };
  message: any;
  error?: {
    message: any;
    stack: any;
  };
}

// ----------------------------------------------------------------------

/**
 * Returns a successful API response with the given status code, response message, and data.
 *
 * @param {Context} c - Hono context object
 * @param {StatusCode} [response_code=200] - HTTP status code
 * @param {string} [response_message="Success!"] - Response message
 * @param {any} message - Response message
 * @param {any} [data=undefined] - Response data
 * @returns {Response} - Hono response object
 */
export const successResponse = (
  c: Context,
  response_code: StatusCode = STATUS_CODES.OK,
  response_message: string = RESPONSE_MESSAGES.SUCCESS,
  message: any,
  data?: any
): Response => {
  c.status(response_code);
  return c.json({
    status: {
      response_code,
      response_message,
    },
    message,
    data,
  } as SuccessResponseData);
};

/**
 * Returns an error API response with the given status code, response message, and error details.
 *
 * @param {Context} c - Hono context object
 * @param {StatusCode} [response_code=400] - HTTP status code
 * @param {string} [response_message="Internal Server Error!"] - Response message
 * @param {any} [error=undefined] - Error details
 * @returns {Response} - Hono response object
 */
export const errorResponse = (
  c: Context,
  response_code: StatusCode = STATUS_CODES.BAD_REQUEST,
  response_message: string = RESPONSE_MESSAGES.SERVER_ERROR,
  error?: any
): Response => {
  c.status(response_code);
  return c.json({
    status: {
      response_code,
      response_message,
    },
    message: error?.message,
    error: error ? { message: error.message, stack: error.stack } : undefined,
  } as ErrorResponseData);
};

File: worker-service/internal/utils/response.go
package utils

// ----------------------------------------------------------------------

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

// ----------------------------------------------------------------------

// Status codes
const (
	StatusOK           = http.StatusOK                  // 200
	StatusCreated      = http.StatusCreated             // 201
	StatusBadRequest   = http.StatusBadRequest          // 400
	StatusUnauthorized = http.StatusUnauthorized        // 401
	StatusForbidden    = http.StatusForbidden           // 403
	StatusNotFound     = http.StatusNotFound            // 404
	StatusConflict     = http.StatusConflict            // 409
	StatusServerError  = http.StatusInternalServerError // 500
)

// Response messages
const (
	ResponseSuccess         = "Success!"
	ResponseBadRequest      = "Bad Request!"
	ResponseUnauthorized    = "Unauthorized!"
	ResponseForbidden       = "Forbidden!"
	ResponseNotFound        = "Not Found!"
	ResponseServerError     = "Internal Server Error!"
	ResponseValidationError = "Validation Error!"
	ResponseResourceExists  = "Resource already exists!"
)

// ----------------------------------------------------------------------

// StatusInfo represents the status object in responses
type StatusInfo struct {
	ResponseCode    int    `json:"response_code"`
	ResponseMessage string `json:"response_message"`
}

// SuccessResponse represents a successful API response
type SuccessResponse struct {
	Status  StatusInfo  `json:"status"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

// ErrorResponse represents an error API response
type ErrorResponse struct {
	Status  StatusInfo   `json:"status"`
	Message string       `json:"message"`
	Error   *ErrorDetail `json:"error,omitempty"`
}

// ErrorDetail contains error details
type ErrorDetail struct {
	Message string `json:"message"`
	Stack   string `json:"stack,omitempty"`
}

// ----------------------------------------------------------------------

// SendSuccessResponse sends a standardized success response
func SendSuccessResponse(c *gin.Context, statusCode int, responseMessage string, message string, data interface{}) {
	c.JSON(statusCode, SuccessResponse{
		Status: StatusInfo{
			ResponseCode:    statusCode,
			ResponseMessage: responseMessage,
		},
		Message: message,
		Data:    data,
	})
}

// SendErrorResponse sends a standardized error response
func SendErrorResponse(c *gin.Context, statusCode int, responseMessage string, err error) {
	errorDetail := &ErrorDetail{
		Message: err.Error(),
	}

	c.JSON(statusCode, ErrorResponse{
		Status: StatusInfo{
			ResponseCode:    statusCode,
			ResponseMessage: responseMessage,
		},
		Message: err.Error(),
		Error:   errorDetail,
	})
}

// ----------------------------------------------------------------------

// Helper functions for common responses
func SuccessOK(c *gin.Context, message string, data interface{}) {
	SendSuccessResponse(c, StatusOK, ResponseSuccess, message, data)
}

func SuccessCreated(c *gin.Context, message string, data interface{}) {
	SendSuccessResponse(c, StatusCreated, ResponseSuccess, message, data)
}

func ErrorBadRequest(c *gin.Context, err error) {
	SendErrorResponse(c, StatusBadRequest, ResponseBadRequest, err)
}

func ErrorUnauthorized(c *gin.Context, err error) {
	SendErrorResponse(c, StatusUnauthorized, ResponseUnauthorized, err)
}

func ErrorNotFound(c *gin.Context, err error) {
	SendErrorResponse(c, StatusNotFound, ResponseNotFound, err)
}

func ErrorConflict(c *gin.Context, err error) {
	SendErrorResponse(c, StatusConflict, ResponseResourceExists, err)
}

func ErrorServerError(c *gin.Context, err error) {
	SendErrorResponse(c, StatusServerError, ResponseServerError, err)
}
